Adds Sass compilation via query parameter

tcaruth opened the pull request:
Enables on-demand Sass compilation by introducing a filter that intercepts requests with a specific query parameter.

+ Supports direct file paths and shorty URLs.
+ Forwards requests to the Sass preprocessor servlet for compilation.

This is a product of a recent multi-site project we're working on. We're using /dA urls to fetch css and javascript from a shared theme, but needed to precompile the css. This addition adds functionality to dotCMS allowing users to specifically request dotsass compilation on a stylesheet, even through a /dA URL.

These should all serve the same file:

- ./application/themes/base/css/styles.dotsass
- ./application/themes/base/css/styles.scss?dotsass=true
- ./dA/{identifier}/styles.scss?dotsass=true

tcaruth committed:
Adds Sass compilation via query parameter

Enables on-demand Sass compilation by introducing a filter that intercepts requests with a specific query parameter.

+ Supports direct file paths and shorty URLs.
+ Forwards requests to the Sass preprocessor servlet for compilation.
+ Includes error handling and logging.

tcaruth commented:
Awaiting a `license.dat` as mentioned by @yolabingo in [dotBackendOnboarding](https://github.com/dotCMS/core/blob/main/dotBackendOnboarding.md#setup)

wezell reviewed (commented):
On thing we ask is that you comment that you agree to the terms of the dotCMS Contributors Agreement found [here](https://gist.github.com/wezell/85ef45298c48494b90d92755b583acb3).  

You can do this by just commenting -  "I agree to the terms of the dotCMS Contributors Licensing Agreement found here: https://gist.github.com/wezell/85ef45298c48494b90d92755b583acb3

tcaruth commented:
I agree to the terms of the dotCMS Contributors Licensing Agreement found here: https://gist.github.com/wezell/85ef45298c48494b90d92755b583acb3

tcaruth committed:
Handles dotsass requests via query parameter

- Processes requests for SCSS files using the `dotsass=true` query
  parameter, supporting both direct file paths and shorty IDs.
- Removes the `SassQueryParamFilter` as the logic is now handled
  within the `CSSPreProcessServlet`.
- Updates the servlet to correctly determine the actual URI of the
  SCSS file, even when using the dotsass parameter with shorty URLs.

tcaruth commented:
Getting close to having this done. I want all these patterns to all work:

- [x] /dA/{identifier}/main.scss?dotsass=true
- [x] /dA/{identifier}?dotsass=true
- [x] /{filename}.dotsass - Unchanged, but tested to ensure no regression
- [ ] /styles.scss?dotsass=true - Still not compiling properly (showing raw SCSS)

I don't find any tests for dotsass in the repo, but maybe I'm not searching correctly. Are there any currently that I can simply add to?

tcaruth committed:
Enhances SASS compilation for shorty URLs

Enables SASS compilation for assets referenced via shorty URLs and direct SCSS files.

- Adds support for the "dotsass=true" parameter when using shorty URLs to trigger SASS preprocessing.
- Forwards requests with "dotsass=true" from ShortyServlet to CSSPreProcessServlet.
- Handles direct SCSS file paths with the "dotsass=true" parameter correctly.
- Improves the logic for identifying and processing SCSS files based on the request URI.

tcaruth committed:
docs: typo in compiler options

tcaruth committed:
Adds filter to handle .scss files with dotsass param

Adds a filter that intercepts requests for .scss files containing the `dotsass=true` query parameter, forwarding them to the CSSPreProcessServlet for SASS compilation.

This ensures that .scss requests are correctly processed even when other servlets might handle them first.
The CSSPreProcessServlet is also updated to check if the request has been forwarded by the new filter.

tcaruth committed:
Merge branch 'main' into feat-dotsass-query-parameter

tcaruth committed:
Changes logging level for forwarded requests

Updates the logging level for messages related to forwarded requests from `info` to `debug` in `CSSPreProcessServlet`, `ScssQueryParamFilter`, and `ShortyServlet`.

This change reduces noise in the logs by only logging these messages at the debug level.

tcaruth committed:
Merge remote-tracking branch 'origin/feat-dotsass-query-parameter' into feat-dotsass-query-parameter

tcaruth commented:
# SASS Query Parameter Feature

## Overview

The SASS Query Parameter feature allows you to compile SCSS files on-the-fly by adding the `dotsass=true` query parameter to URLs. This feature extends the existing SASS compilation capabilities in dotCMS to support additional URL patterns.

## Supported URL Patterns

The following URL patterns now support SASS compilation with the `dotsass=true` query parameter:

1. **Direct SCSS file access**:
   ```
   /path/to/styles.scss?dotsass=true
   ```

2. **Shorty URLs** (file assets accessed via `/dA/` path):
   ```
   /dA/{identifier}?dotsass=true
   ```

3. **Traditional SASS compilation** (still supported):
   ```
   /path/to/styles.dotsass
   ```

## Implementation Details

The feature is implemented through:

1. A filter (`ScssQueryParamFilter`) that intercepts `.scss` requests with the `dotsass=true` parameter and forwards them to the SASS compiler
2. Request handling in `CSSPreProcessServlet` to process forwarded requests
3. Integration with `ShortyServlet` to handle `/dA/` URLs with the `dotsass=true` parameter

## Use Cases

### Frontend Development

When working with SCSS files in dotCMS, you can now preview the compiled CSS by simply adding `?dotsass=true` to the URL of any SCSS file, without having to modify the file path or extension.

### Theme Development

For theme developers, this feature allows for easier testing of SCSS changes by enabling direct access to the compiled CSS through the original SCSS file URL with the query parameter.

This also eases multi-site shared theme development by adding SASS compilation via /dA/ urls. This ensures there are no issues with cross-host stylesheets and no need to make the shared theme directly available via a URL. This also helps with multi-site shared theme implementations running locally (typically for development purposes)

### Content Delivery

The feature also supports compiled SCSS delivery through the `/dA/` shorty URLs, enabling more flexible content delivery patterns while maintaining SASS compilation capabilities.

## Performance Considerations

- The feature only intercepts requests with the specific `dotsass=true` parameter, ensuring minimal impact on other requests

## Example Usage

```html
<!-- Regular CSS inclusion -->
<link rel="stylesheet" href="/path/to/styles.css">

<!-- SCSS with on-the-fly compilation -->
<link rel="stylesheet" href="/path/to/styles.scss?dotsass=true">

<!-- SCSS via shorty URL with on-the-fly compilation -->
<link rel="stylesheet" href="/dA/12345?dotsass=true">
```

## Related Documentation

- [SASS Compiler Documentation](https://dev.dotcms.com/docs/sass-compiler)

wezell commented on dotCMS/src/main/java/com/dotcms/csspreproc/CSSPreProcessServlet.java:
```diff
@@ -61,13 +64,68 @@ public void doGet(final HttpServletRequest req, final HttpServletResponse resp)
+            
+            // Check for dotsass=true query parameter to enable SASS compilation for non-standard URLs
+            final String dotsassParam = req.getParameter("dotsass");
+            final boolean isDotsassParam = dotsassParam != null && dotsassParam.equalsIgnoreCase("true");
```
could be shortened to `"true".equalsIgnoreCase(dotsassParam)`

tcaruth replied:
good catch! I'll shorten both of these in the next commit

wezell commented on dotCMS/src/main/java/com/dotcms/csspreproc/ScssQueryParamFilter.java:
```diff
@@ -0,0 +1,55 @@
+        
+        String uri = httpRequest.getRequestURI();
+        String dotsassParam = request.getParameter("dotsass");
+        boolean isDotsassParam = dotsassParam != null && dotsassParam.equalsIgnoreCase("true");
```
same as above, could be shorter

wezell reviewed (changes requested):
Take a look Travis, let me know your thoughts

tcaruth committed:
Simplifies boolean check for dotsass parameter

Refactors the way the "dotsass" request parameter is checked for equality to "true" based on feedback from @wezell
https://github.com/dotCMS/core/pull/32885#pullrequestreview-3088971155

tcaruth committed:
Handles file asset retrieval based on short type (inode and identifier)

Ensures the correct retrieval of file assets by checking the short type.
Now uses the appropriate API call based on whether the short type is an identifier or inode, ensuring that file assets are correctly fetched.

https://github.com/dotCMS/core/pull/32885#pullrequestreview-3088998174

tcaruth committed:
Revert: reorder filter mappings in web.xml

Reverts https://github.com/dotCMS/core/blob/353bbe70b328fdf9bf2a97dea05461754a59f9ef/dotCMS/src/main/webapp/WEB-INF/web.xml
